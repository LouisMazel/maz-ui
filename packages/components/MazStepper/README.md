# MazStepper

## Props

<!-- @vuese:MazStepper:props:start -->

| Name                  | Description                               | Type      | Required | Default |
| --------------------- | ----------------------------------------- | --------- | -------- | ------- |
| value                 | value of active step                      | `Number`  | `true`   | -       |
| steps                 | steps number                              | `Number`  | `false`  | 1       |
| color                 | choose a color from list                  | `String`  | `false`  | primary |
| showStepNumber        | If is `true`, item number is shown        | `Boolean` | `false`  | false   |
| space                 | flexbox space (`around`, `between`)       | `String`  | `false`  | around  |
| size                  | step size                                 | `Number`  | `false`  | 10      |
| variant               | step style (`dot`, `square`)              | `String`  | `false`  | dot     |
| shadow                | add shadow elevation to step buttons      | `Boolean` | `false`  | false   |
| disabled              | disallow step click                       | `Boolean` | `false`  | false   |
| disabledSteps         | disallow multiple steps click             | `Array`   | `false`  | Array   |
| disabledNextSteps     | disallow next steps from current step     | `Boolean` | `false`  | false   |
| disabledPreviousSteps | disallow previous steps from current step | `Boolean` | `false`  | false   |
| dark                  | set dark mode                             | `Boolean` | `false`  | false   |

<!-- @vuese:MazStepper:props:end -->

## Events

<!-- @vuese:MazStepper:events:start -->

| Event Name | Description                    | Parameters |
| ---------- | ------------------------------ | ---------- |
| input      | return the step number clicked | `Number`   |

<!-- @vuese:MazStepper:events:end -->
