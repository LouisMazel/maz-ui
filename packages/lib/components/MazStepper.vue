<template>
  <div class="m-stepper">
    <template v-for="step in stepCount" :key="step">
      <button
        v-if="hasDataForStep('title', step)"
        type="button"
        :disabled="isStepDisabled(step)"
        class="m-stepper__header"
        :class="[
          {
            '--is-current-step': step === currentStep || allStepsOpened,
          },
          `${getStepStateData(step).class}`,
        ]"
        @click="selectStep(step)"
      >
        <div class="m-stepper__header__wrapper">
          <span class="m-stepper__count --primary">
            <div class="m-stepper__count__circle">
              <Component
                :is="getStepStateData(step).icon"
                v-if="getStepStateData(step).icon"
                class="icon maz-text-xl"
              />
            </div>

            {{ step }}
          </span>

          <div class="m-stepper__header__content">
            <span class="m-stepper__title">
              <slot :name="`title-${step}`">
                {{ getPropertyInStep('title', step) }}
              </slot>
            </span>
            <span v-if="hasDataForStep('subtitle', step)" class="m-stepper__subtitle">
              <slot :name="`subtitle-${step}`">
                {{ getPropertyInStep('subtitle', step) }}
              </slot>
            </span>
          </div>
        </div>

        <span v-if="hasDataForStep('titleInfo', step)" class="m-stepper__right">
          <slot :name="`title-info-${step}`">
            {{ getPropertyInStep('titleInfo', step) }}
          </slot>
        </span>
      </button>

      <div
        class="m-stepper__content"
        :class="{
          '--no-border': isLastStep(step),
        }"
      >
        <MazTransitionExpand>
          <div v-show="allStepsOpened || currentStep === step">
            <div class="m-stepper__content__wrapper">
              <slot
                :name="`content-${step}`"
                :validated="isStepSuccess(step)"
                :error="isStepError(step)"
                :warning="isStepWarning(step)"
                :next-step="() => selectStep(step + 1)"
                :previous-step="() => selectStep(step - 1)"
              ></slot>
            </div>
          </div>
        </MazTransitionExpand>
      </div>
    </template>
  </div>
</template>

<script lang="ts" setup>
  import { computed, useSlots, ref, type PropType, defineAsyncComponent, type Component } from 'vue'
  import type { Color } from './types'

  export type { Color }

  export interface Step {
    title?: string
    subtitle?: string
    titleInfo?: string
    disabled?: boolean
    error?: boolean
    success?: boolean
    warning?: boolean
  }
  export type Steps = Step[]

  const MazTransitionExpand = defineAsyncComponent(() => import('./MazTransitionExpand.vue'))
  const CheckCircleIcon = defineAsyncComponent(() => import('./../icons/check-circle.svg'))
  const ExclamationCircleIcon = defineAsyncComponent(
    () => import('./../icons/exclamation-circle.svg'),
  )
  const ExclamationIcon = defineAsyncComponent(() => import('./../icons/exclamation-triangle.svg'))

  const props = defineProps({
    modelValue: { type: Number, default: undefined },
    steps: { type: Array as PropType<Steps>, default: undefined },
    disabledNextSteps: { type: Boolean, default: false },
    disabledPreviousSteps: { type: Boolean, default: false },
    autoValidateSteps: { type: Boolean, default: false },
    allStepsOpened: { type: Boolean, default: false },
    allStepsValidated: { type: Boolean, default: false },
    color: {
      type: String as PropType<Color>,
      default: 'primary',
    },
    canCloseSteps: { type: Boolean, default: false },
  })

  const roundStepBgColor = computed(() => `var(--maz-color-${props.color})`)
  const roundStepTextColor = computed(() => `var(--maz-color-${props.color}-contrast)`)

  const emits = defineEmits(['update:model-value'])

  const slots = useSlots()

  const stepCount = computed<number>(
    () => Object.keys(slots).filter((slot) => slot.startsWith('content-')).length,
  )

  const localModelValue = ref(1)

  const currentStep = computed({
    get: () => props.modelValue ?? localModelValue.value,
    set: (value: number) => {
      localModelValue.value = value
      emits('update:model-value', value)
    },
  })

  const getStepStateData = (step: number): { icon?: Component; class: string } => {
    if (isStepSuccess(step)) {
      return { icon: CheckCircleIcon, class: '--success' }
    } else if (isStepWarning(step)) {
      return { icon: ExclamationIcon, class: '--warning' }
    } else if (isStepError(step)) {
      return { icon: ExclamationCircleIcon, class: '--error' }
    }

    return { class: '--normal' }
  }

  const getPropertyInStep = (property: 'title' | 'titleInfo' | 'subtitle', step: number) => {
    return props.steps?.[step - 1]?.[property]
  }

  const selectStep = (step: number) => {
    if (currentStep.value === step && props.canCloseSteps) {
      currentStep.value = 0
    } else if (step < 1) {
      currentStep.value = 1
    } else if (step > stepCount.value) {
      currentStep.value = stepCount.value
    } else {
      currentStep.value = step
    }
  }

  const hasDataForStep = (property: 'title' | 'titleInfo' | 'subtitle', step: number): boolean => {
    const data = property === 'titleInfo' ? 'title-info' : property

    const hasSlot = Object.keys(slots)
      .filter((slot) => slot.startsWith(`${data}-`))
      .includes(`${data}-${step}`)

    const hasData = !!props.steps?.[step - 1]?.[property]

    return hasSlot || hasData
  }

  const isStepSuccess = (step: number): boolean => {
    const isValidated = props.steps?.[step - 1]?.success
    const hasErrorOrWarningState = isStepError(step) || isStepWarning(step)

    const isAutoValidated =
      props.autoValidateSteps && step < currentStep.value && !hasErrorOrWarningState
    return isValidated ?? (isAutoValidated || props.allStepsValidated)
  }

  const isStepDisabled = (step: number): boolean => {
    const isDisabled = props.steps?.[step - 1]?.disabled
    const isCurrentStepDisabled = currentStep.value === step && !props.canCloseSteps
    const isAutoDisabledNext = props.disabledNextSteps && step > currentStep.value
    const isAutoDisabledPrevious = props.disabledPreviousSteps && step < currentStep.value

    return (
      isDisabled ??
      (isCurrentStepDisabled ||
        isAutoDisabledNext ||
        isAutoDisabledPrevious ||
        props.allStepsOpened)
    )
  }

  const isStepError = (step: number) => {
    return props.steps?.[step - 1]?.error
  }

  const isStepWarning = (step: number) => {
    return props.steps?.[step - 1]?.warning
  }

  const isLastStep = (step: number): boolean => {
    return step === stepCount.value
  }
</script>

<style lang="postcss" scoped>
  .m-stepper {
    &__header {
      @apply maz-flex maz-w-full maz-cursor-pointer maz-select-text maz-items-center
        maz-justify-between maz-space-x-4 maz-rounded maz-px-4 maz-py-2 maz-outline-none
        maz-transition-colors maz-duration-200;

      &__content {
        @apply maz-flex maz-flex-col maz-items-start;
      }

      &:not(:disabled) {
        @apply hover:maz-bg-color-lighter focus:maz-bg-color-lighter;
      }

      &:disabled {
        @apply maz-cursor-not-allowed;
      }

      &.--is-current-step {
        @apply maz-cursor-default;
      }

      &__wrapper {
        @apply maz-flex maz-items-center maz-space-x-4;
      }

      &.--success {
        .m-stepper__count__circle {
          @apply maz-scale-100 maz-bg-success;
        }

        .m-stepper__right {
          @apply maz-text-success;
        }

        svg {
          @apply maz-text-success-contrast;
        }
      }

      &.--warning {
        .m-stepper__count__circle {
          @apply maz-scale-100 maz-bg-warning;
        }

        .m-stepper__right {
          @apply maz-text-warning;
        }

        svg {
          @apply maz-text-warning-contrast;
        }
      }

      &.--error {
        .m-stepper__count__circle {
          @apply maz-scale-100 maz-bg-danger;
        }

        .m-stepper__right {
          @apply maz-text-danger;
        }

        svg {
          @apply maz-text-danger-contrast;
        }
      }
    }

    &__title {
      @apply maz-text-normal;
    }

    &__subtitle {
      @apply maz-mt-1 maz-text-xs maz-text-muted;
    }

    &__right {
      @apply maz-text-right maz-text-sm maz-text-primary;
    }

    &__count {
      @apply maz-relative maz-flex maz-h-8 maz-w-8 maz-flex-none
        maz-overflow-hidden maz-rounded-full maz-text-lg maz-flex-center;

      background-color: v-bind('roundStepBgColor');
      color: v-bind('roundStepTextColor');

      &__circle {
        @apply maz-absolute maz-inset-0 maz-flex maz-scale-0 maz-rounded-full
          maz-transition-all maz-duration-300 maz-ease-in-out maz-flex-center;
      }

      svg {
        @apply maz-text-success-contrast;
      }
    }

    &__content {
      @apply maz-ml-[2rem] maz-border-l-2 maz-border-transparent maz-py-2 maz-pl-8;

      &__wrapper {
        @apply maz-py-2;
      }

      &:not(.--no-border) {
        @apply maz-border-color-lighter;
      }
    }
  }
</style>
