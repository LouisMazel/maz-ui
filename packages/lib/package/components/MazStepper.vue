<template>
  <div class="m-stepper">
    <template v-for="step in stepCount" :key="step">
      <button
        v-if="hasTitleForStep(step)"
        type="button"
        :disabled="isStepDisabled(step)"
        class="m-stepper__header"
        @click="selectStep(step)"
      >
        <span
          class="m-stepper__count --primary"
          :class="{
            '--validated': isStepValidated(step),
          }"
        >
          <svg
            v-if="isStepValidated(step)"
            xmlns="http://www.w3.org/2000/svg"
            fill="none"
            viewBox="0 0 24 24"
            width="1.2rem"
            height="1.2rem"
            stroke="currentColor"
            aria-hidden="true"
          >
            <path
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              d="M5 13l4 4L19 7"
            />
          </svg>

          <template v-else>
            {{ step }}
          </template>
        </span>

        <div class="m-stepper__header__content">
          <span class="m-stepper__title">
            <slot :name="`title-${step}`" />
          </span>
          <span v-if="hasSubtitleForStep(step)" class="m-stepper__subtitle">
            <slot :name="`subtitle-${step}`" />
          </span>
        </div>
      </button>

      <div
        class="m-stepper__content"
        :class="{
          '--no-border': isLastStep(step),
        }"
      >
        <MazTransitionExpand>
          <div v-if="currentStep === step">
            <div class="m-stepper__content__wrapper">
              <slot
                :name="`content-${step}`"
                :next-step="() => selectStep(step + 1)"
                :previous-step="() => selectStep(step - 1)"
              />
            </div>
          </div>
        </MazTransitionExpand>
      </div>
    </template>
  </div>
</template>

<script lang="ts" setup>
  import { computed, useSlots, ref, type PropType } from 'vue'
  import MazTransitionExpand from './MazTransitionExpand.vue'

  export interface Step {
    disabled?: boolean
    validated?: boolean
  }

  const props = defineProps({
    modelValue: { type: Number, default: undefined },
    steps: { type: Array as PropType<Step[]>, default: undefined },
    disabledNextSteps: { type: Boolean, default: false },
    disabledPreviousSteps: { type: Boolean, default: false },
    autoValidatedSteps: { type: Boolean, default: false },
  })

  const emits = defineEmits(['update:model-value'])

  const slots = useSlots()

  const stepCount = computed<number>(
    () =>
      Object.keys(slots).filter((slot) => slot.startsWith('content-')).length,
  )

  const localModelValue = ref(1)

  const currentStep = computed({
    get: () => props.modelValue ?? localModelValue.value,
    set: (value: number) => {
      localModelValue.value = value
      emits('update:model-value', value)
    },
  })

  const selectStep = (step: number) => {
    if (step < 1) {
      currentStep.value = 1
    } else if (step > stepCount.value) {
      currentStep.value = stepCount.value
    } else {
      currentStep.value = step
    }
  }

  const hasTitleForStep = (step: number): boolean => {
    return Object.keys(slots)
      .filter((slot) => slot.startsWith('title-'))
      .includes(`title-${step}`)
  }

  const hasSubtitleForStep = (step: number): boolean => {
    return Object.keys(slots)
      .filter((slot) => slot.startsWith('subtitle-'))
      .includes(`subtitle-${step}`)
  }

  const isStepValidated = (step: number): boolean => {
    const isValidated = !!props.steps?.[step - 1]?.validated
    const isAutoValidated = props.autoValidatedSteps && step < currentStep.value
    return isValidated || isAutoValidated
  }

  const isStepDisabled = (step: number): boolean => {
    const isDisabled = !!props.steps?.[step - 1]?.disabled
    const isCurrentStep = currentStep.value === step
    const isAutoDisabledNext =
      props.disabledNextSteps && step > currentStep.value
    const isAutoDisabledPrevious =
      props.disabledPreviousSteps && step < currentStep.value

    return (
      isDisabled ||
      isCurrentStep ||
      isAutoDisabledNext ||
      isAutoDisabledPrevious
    )
  }

  const isLastStep = (step: number): boolean => {
    return step === stepCount.value
  }
</script>

<style lang="postcss" scoped>
  .m-stepper {
    &__header {
      @apply maz-flex maz-w-full maz-cursor-pointer maz-items-center maz-space-x-4
        maz-rounded-lg maz-px-4 maz-py-2 maz-outline-none maz-transition-colors maz-duration-200;

      &__content {
        @apply maz-flex maz-flex-col maz-items-start;
      }

      &:not(:disabled) {
        @apply hover:maz-bg-color-lighter focus:maz-bg-color-lighter;
      }

      &:disabled {
        @apply maz-cursor-not-allowed;
      }
    }

    &__title {
      @apply maz-text-normal;
    }

    &__subtitle {
      @apply maz-text-xs maz-text-muted;
    }

    &__count {
      @apply maz-flex maz-h-8 maz-w-8 maz-flex-none maz-items-center maz-justify-center maz-rounded-full maz-text-lg;

      &.--primary {
        @apply maz-bg-primary maz-text-primary-contrast;
      }
    }

    &__content {
      @apply maz-ml-[2rem] maz-border-l-2 maz-border-transparent maz-py-2 maz-pl-8;

      &__wrapper {
        @apply maz-py-2;
      }

      &:not(.--no-border) {
        @apply maz-border-color-lighter;
      }
    }
  }
</style>
