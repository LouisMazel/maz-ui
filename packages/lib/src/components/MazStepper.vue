<script lang="ts" setup>
import type { IconComponent } from '@maz-ui/icons'
import type { Component } from 'vue'
import type { MazColor } from './types'
import { MazCheckCircle, MazExclamationCircle, MazExclamationTriangle } from '@maz-ui/icons'
import {
  computed,
  ref,
  useSlots,
} from 'vue'

import MazExpandAnimation from './MazExpandAnimation.vue'
import MazIcon from './MazIcon.vue'

export interface MazStepperStep {
  title?: string
  subtitle?: string
  titleInfo?: string
  disabled?: boolean
  error?: boolean
  success?: boolean
  warning?: boolean
  icon?: string | IconComponent
}

const {
  modelValue,
  steps,
  color = 'primary',
  disabledNextSteps,
  disabledPreviousSteps,
  autoValidateSteps,
  allStepsOpened,
  allStepsValidated,
  canCloseSteps,
} = defineProps<MazStepperProps>()

const emits = defineEmits<{
  'update:model-value': [value: number]
}>()

export interface MazStepperProps {
  /** The current step */
  modelValue?: number
  /** The steps */
  steps?: MazStepperStep[]
  /**
   * The color of the stepper
   * @default primary
   */
  color?: MazColor
  /** Disable the next steps */
  disabledNextSteps?: boolean
  /** Disable the previous steps */
  disabledPreviousSteps?: boolean
  /** Auto validate the steps */
  autoValidateSteps?: boolean
  /** Open all steps */
  allStepsOpened?: boolean
  /** Validate all steps */
  allStepsValidated?: boolean
  /** Allow to close the steps */
  canCloseSteps?: boolean
}

const roundStepBgColor = computed(() => `hsl(var(--maz-${color}))`)
const roundStepTextColor = computed(() => `hsl(var(--maz-${color}-foreground))`)

const slots = useSlots()

const stepCount = computed<number>(
  () => Object.keys(slots).filter(slot => slot.startsWith('content-')).length,
)

const localModelValue = ref(1)

const currentStep = computed({
  get: () => modelValue ?? localModelValue.value,
  set: (value: number) => {
    localModelValue.value = value
    emits('update:model-value', value)
  },
})

function getStepStateData(step: number): { icon?: Component, class: string } {
  if (isStepSuccess(step)) {
    return { icon: MazCheckCircle, class: '--success' }
  }
  else if (isStepWarning(step)) {
    return { icon: MazExclamationTriangle, class: '--warning' }
  }
  else if (isStepError(step)) {
    return { icon: MazExclamationCircle, class: '--error' }
  }

  return { class: '--normal' }
}

function getStepIcon(step: number): IconComponent | string | undefined {
  return steps?.[step - 1]?.icon
}

function getPropertyInStep(property: 'title' | 'titleInfo' | 'subtitle', step: number) {
  return steps?.[step - 1]?.[property]
}

function selectStep(step: number) {
  if (currentStep.value === step && canCloseSteps) {
    currentStep.value = 0
  }
  else if (step < 1) {
    currentStep.value = 1
  }
  else if (step > stepCount.value) {
    currentStep.value = stepCount.value
  }
  else {
    currentStep.value = step
  }
}

function hasDataForStep(property: 'title' | 'titleInfo' | 'subtitle', step: number): boolean {
  const data = property === 'titleInfo' ? 'title-info' : property

  const hasSlot = Object.keys(slots)
    .filter(slot => slot.startsWith(`${data}-`))
    .includes(`${data}-${step}`)

  const hasData = !!steps?.[step - 1]?.[property]

  return hasSlot || hasData
}

function isStepSuccess(step: number): boolean {
  const isValidated = steps?.[step - 1]?.success
  const hasErrorOrWarningState = isStepError(step) || isStepWarning(step)

  const isAutoValidated
      = autoValidateSteps && step < currentStep.value && !hasErrorOrWarningState
  return isValidated ?? (isAutoValidated || allStepsValidated)
}

function isStepDisabled(step: number): boolean {
  const isDisabled = steps?.[step - 1]?.disabled
  const isCurrentStepDisabled = currentStep.value === step && !canCloseSteps
  const isAutoDisabledNext = disabledNextSteps && step > currentStep.value
  const isAutoDisabledPrevious = disabledPreviousSteps && step < currentStep.value

  return (
    isDisabled
    ?? (isCurrentStepDisabled
      || isAutoDisabledNext
      || isAutoDisabledPrevious
      || allStepsOpened)
  )
}

function isStepError(step: number) {
  return steps?.[step - 1]?.error
}

function isStepWarning(step: number) {
  return steps?.[step - 1]?.warning
}

function isLastStep(step: number): boolean {
  return step === stepCount.value
}
</script>

<template>
  <div class="m-stepper m-reset-css" :style="[{ '--round-step-bg-color': roundStepBgColor, '--round-step-text-color': roundStepTextColor }]">
    <template v-for="step in stepCount" :key="step">
      <button
        v-if="hasDataForStep('title', step)"
        :id="`header-step-${step}`"
        type="button"
        :disabled="isStepDisabled(step)"
        class="m-stepper__header"
        :class="[
          {
            '--is-current-step': step === currentStep || allStepsOpened,
            '--disabled': step !== currentStep && !allStepsOpened && isStepDisabled(step),
          },
          `${getStepStateData(step).class}`,
        ]"
        @click="selectStep(step)"
      >
        <div class="m-stepper__header__wrapper">
          <span class="m-stepper__count --primary">
            <div class="m-stepper__count__circle">
              <Component
                :is="getStepStateData(step).icon"
                v-if="getStepStateData(step).icon"
                class="icon maz-text-xl"
              />
            </div>
            <!--
              @slot icon-${step} - Replace step number in the circle by an icon for the step
            -->
            <slot :name="`icon-${step}`">
              <template v-if="getStepIcon(step)">
                <MazIcon
                  v-if="typeof getStepIcon(step) === 'string'"
                  :name="getStepIcon(step) as string"
                />
                <Component :is="getStepIcon(step)" v-else-if="getStepIcon(step)" />
              </template>
              <template v-else>
                {{ step }}
              </template>
            </slot>
          </span>

          <div class="m-stepper__header__content">
            <span class="m-stepper__title">
              <!--
                @slot title-${step} - Title of the step
              -->
              <slot :name="`title-${step}`">
                {{ getPropertyInStep('title', step) }}
              </slot>
            </span>
            <span v-if="hasDataForStep('subtitle', step)" class="m-stepper__subtitle">
              <!--
                @slot title-${step} - Subtitle of the step
              -->
              <slot :name="`subtitle-${step}`">
                {{ getPropertyInStep('subtitle', step) }}
              </slot>
            </span>
          </div>
        </div>

        <span v-if="hasDataForStep('titleInfo', step)" class="m-stepper__right">
          <!--
            @slot title-info-${step} - Info of the right of the step
          -->
          <slot :name="`title-info-${step}`">
            {{ getPropertyInStep('titleInfo', step) }}
          </slot>
        </span>
      </button>

      <div
        class="m-stepper__content"
        :class="{
          '--no-border': isLastStep(step),
        }"
      >
        <MazExpandAnimation
          :model-value="allStepsOpened || currentStep === step"
          :aria-labelledby="`header-step-${step}`"
        >
          <div class="m-stepper__content__wrapper">
            <!-- @slot content-${step} - Content of the step
                @binding {boolean} validated - If the step is validated
                @binding {boolean} error - If the step has an error
                @binding {boolean} warning - If the step has a warning
                @binding {Function} previous-step - Function to go to the previous step
                @binding {Function} next-step - Function to go to the next step
              -->
            <slot
              :name="`content-${step}`"
              :validated="isStepSuccess(step)"
              :error="isStepError(step)"
              :warning="isStepWarning(step)"
              :next-step="() => selectStep(step + 1)"
              :previous-step="() => selectStep(step - 1)"
            />
          </div>
        </MazExpandAnimation>
      </div>
    </template>
  </div>
</template>

<style lang="postcss" scoped>
  .m-stepper {
  &__right {
    @apply maz-truncate maz-text-end maz-text-sm maz-text-primary;
  }

  &__header {
    @apply maz-flex maz-w-full maz-cursor-pointer maz-select-text maz-items-center
        maz-justify-between maz-space-x-4 maz-rounded maz-px-4 maz-py-2 maz-text-start
        maz-transition-colors maz-duration-200;

    &__content {
      @apply maz-flex maz-flex-none maz-flex-col maz-items-start;
    }

    &:not(:disabled) {
      @apply hover:maz-bg-surface-300 hover:dark:maz-bg-surface-400;
    }

    &:disabled {
      @apply maz-cursor-not-allowed;
    }

    &.--disabled {
      @apply maz-text-gray-400 dark:maz-text-gray-500;
    }

    &.--is-current-step {
      @apply maz-cursor-default;
    }

    &__wrapper {
      @apply maz-flex maz-items-center maz-space-x-4;
    }

    &.--success {
      .m-stepper__count__circle {
        @apply maz-scale-100 maz-bg-success;
      }

      .m-stepper__right {
        @apply maz-text-success;
      }

      svg {
        @apply maz-text-success-foreground;
      }
    }

    &.--warning {
      .m-stepper__count__circle {
        @apply maz-scale-100 maz-bg-warning;
      }

      .m-stepper__right {
        @apply maz-text-warning;
      }

      svg {
        @apply maz-text-warning-foreground;
      }
    }

    &.--error {
      .m-stepper__count__circle {
        @apply maz-scale-100 maz-bg-destructive;
      }

      .m-stepper__right {
        @apply maz-text-destructive;
      }

      svg {
        @apply maz-text-destructive-foreground;
      }
    }
  }

  &__title {
    @apply maz-text-lg;
  }

  &__subtitle {
    @apply maz-mt-1 maz-text-sm maz-text-muted;
  }

  &__count {
    @apply maz-relative maz-flex maz-h-8 maz-w-8 maz-flex-none
        maz-overflow-hidden maz-rounded-full maz-text-lg maz-flex-center;

    background-color: var(--round-step-bg-color);
    color: var(--round-step-text-color);

    &__circle {
      @apply maz-absolute maz-inset-0 maz-flex maz-scale-0 maz-rounded-full
          maz-transition-all maz-duration-300 maz-ease-in-out maz-flex-center;
    }

    svg {
      @apply maz-text-success-foreground;
    }
  }

  &__content {
    @apply maz-ml-[1.95rem] maz-border-l-2 maz-border-transparent maz-py-2 maz-pl-8;

    &__wrapper {
      @apply maz-py-2;
    }

    &:not(.--no-border) {
      @apply maz-border-divider;
    }
  }
}
</style>
