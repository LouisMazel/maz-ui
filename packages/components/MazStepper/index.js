import MazStepper from './_main.vue'

MazStepper.install = (Vue) => {
  Vue.component(MazStepper.name, MazStepper)
}

export default MazStepper
