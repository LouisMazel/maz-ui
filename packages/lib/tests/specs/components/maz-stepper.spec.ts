import type { VueWrapper } from '@vue/test-utils'
import { shallowMount } from '@vue/test-utils'
import MazStepper from '@components/MazStepper.vue'
import type { ComponentPublicInstance } from 'vue'

describe('components/MazStepper.vue', () => {
  expect(MazStepper).toBeTruthy()

  let wrapper: VueWrapper<ComponentPublicInstance & { [key: string]: any }>

  beforeEach(() => {
    wrapper = shallowMount(MazStepper, {
      props: {
        modelValue: 1,
        steps: [{ success: true, disabled: true }],
      },
      slots: {
        'title-1': 'Step 1',
        'content-1': 'Content 1',
        'subtitle-2': 'subtitle 2',
        'content-2': 'Content 2',
        'title-3': 'Step 3',
        'content-3': 'Content 3',
      },
    })
  })

  test('Should match with the snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot()
  })

  test('Should have the current step according with the model', () => {
    expect(wrapper.vm.stepCount).toBe(3)
    expect(wrapper.vm.localModelValue).toBe(1)
    expect(wrapper.vm.currentStep).toBe(1)
  })

  test('Should have the current step according with the model', () => {
    wrapper.vm.selectStep(2)
    expect(wrapper.vm.localModelValue).toBe(2)
    expect(wrapper.emitted()['update:model-value']).toStrictEqual([[2]])
  })

  test('Should returns if title exists', () => {
    const title1 = wrapper.vm.hasDataForStep('title', 1)
    expect(title1).toBeTruthy()
    const title2 = wrapper.vm.hasDataForStep('title', 2)
    expect(title2).toBeFalsy()
  })

  test('Should returns if subtitle exists', () => {
    const subtitle1 = wrapper.vm.hasDataForStep('subtitle', 1)
    expect(subtitle1).toBeFalsy()
    const subtitle2 = wrapper.vm.hasDataForStep('subtitle', 2)
    expect(subtitle2).toBeTruthy()
  })

  test('Should returns if step is validated', () => {
    const validated1 = wrapper.vm.isStepSuccess(1)
    expect(validated1).toBeTruthy()
    const validated2 = wrapper.vm.isStepSuccess(2)
    expect(validated2).toBeFalsy()
  })

  test('Should returns if step is disabled', () => {
    const disabled1 = wrapper.vm.isStepDisabled(1)
    expect(disabled1).toBeTruthy()
    const disabled2 = wrapper.vm.isStepDisabled(2)
    expect(disabled2).toBeFalsy()
  })

  test('Should returns if step is the last', () => {
    const isLast1 = wrapper.vm.isLastStep(1)
    expect(isLast1).toBeFalsy()
    const isLast2 = wrapper.vm.isLastStep(3)
    expect(isLast2).toBeTruthy()
  })

  test('Should open step if click on it', async () => {
    const buttons = wrapper.findAll('button')
    await buttons[1].trigger('click')
    expect(wrapper.emitted()['update:model-value']).toStrictEqual([[3]])
  })
})
