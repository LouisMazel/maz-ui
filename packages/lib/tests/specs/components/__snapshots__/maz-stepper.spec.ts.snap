// Vitest Snapshot v1

exports[`components/MazStepper.vue > Should match with the snapshot 1`] = `
"<div class=\\"m-stepper\\" data-v-8231b20c=\\"\\"><button type=\\"button\\" disabled=\\"\\" class=\\"m-stepper__header --is-current-step\\" data-v-8231b20c=\\"\\">
    <div class=\\"m-stepper__header__wrapper\\" data-v-8231b20c=\\"\\"><span class=\\"m-stepper__count --primary --validated\\" data-v-8231b20c=\\"\\"><div class=\\"m-stepper__count__circle\\" data-v-8231b20c=\\"\\"><svg xmlns=\\"http://www.w3.org/2000/svg\\" fill=\\"none\\" viewBox=\\"0 0 24 24\\" width=\\"1.2rem\\" height=\\"1.2rem\\" stroke=\\"currentColor\\" aria-hidden=\\"true\\" data-v-8231b20c=\\"\\"><path stroke-linecap=\\"round\\" stroke-linejoin=\\"round\\" stroke-width=\\"2\\" d=\\"M5 13l4 4L19 7\\" data-v-8231b20c=\\"\\"></path></svg></div> 1</span>
      <div class=\\"m-stepper__header__content\\" data-v-8231b20c=\\"\\"><span class=\\"m-stepper__title\\" data-v-8231b20c=\\"\\">Step 1</span>
        <!--v-if-->
      </div>
    </div>
    <!--v-if-->
  </button>
  <div class=\\"m-stepper__content\\" data-v-8231b20c=\\"\\">
    <transition-group-stub name=\\"maz-expand\\" tag=\\"div\\" appear=\\"false\\" persisted=\\"false\\" css=\\"true\\" class=\\"m-transition-expand\\" data-v-f61b1628=\\"\\" data-v-8231b20c=\\"\\">
      <div data-v-8231b20c=\\"\\">
        <div class=\\"m-stepper__content__wrapper\\" data-v-8231b20c=\\"\\">Content 1</div>
      </div>
    </transition-group-stub>
  </div>
  <!--v-if-->
  <div class=\\"m-stepper__content\\" data-v-8231b20c=\\"\\">
    <transition-group-stub name=\\"maz-expand\\" tag=\\"div\\" appear=\\"false\\" persisted=\\"false\\" css=\\"true\\" class=\\"m-transition-expand\\" data-v-f61b1628=\\"\\" data-v-8231b20c=\\"\\">
      <div data-v-8231b20c=\\"\\" style=\\"display: none;\\">
        <div class=\\"m-stepper__content__wrapper\\" data-v-8231b20c=\\"\\">Content 2</div>
      </div>
    </transition-group-stub>
  </div><button type=\\"button\\" class=\\"m-stepper__header\\" data-v-8231b20c=\\"\\">
    <div class=\\"m-stepper__header__wrapper\\" data-v-8231b20c=\\"\\"><span class=\\"m-stepper__count --primary\\" data-v-8231b20c=\\"\\"><div class=\\"m-stepper__count__circle\\" data-v-8231b20c=\\"\\"><svg xmlns=\\"http://www.w3.org/2000/svg\\" fill=\\"none\\" viewBox=\\"0 0 24 24\\" width=\\"1.2rem\\" height=\\"1.2rem\\" stroke=\\"currentColor\\" aria-hidden=\\"true\\" data-v-8231b20c=\\"\\"><path stroke-linecap=\\"round\\" stroke-linejoin=\\"round\\" stroke-width=\\"2\\" d=\\"M5 13l4 4L19 7\\" data-v-8231b20c=\\"\\"></path></svg></div> 3</span>
      <div class=\\"m-stepper__header__content\\" data-v-8231b20c=\\"\\"><span class=\\"m-stepper__title\\" data-v-8231b20c=\\"\\">Step 3</span>
        <!--v-if-->
      </div>
    </div>
    <!--v-if-->
  </button>
  <div class=\\"m-stepper__content --no-border\\" data-v-8231b20c=\\"\\">
    <transition-group-stub name=\\"maz-expand\\" tag=\\"div\\" appear=\\"false\\" persisted=\\"false\\" css=\\"true\\" class=\\"m-transition-expand\\" data-v-f61b1628=\\"\\" data-v-8231b20c=\\"\\">
      <div data-v-8231b20c=\\"\\" style=\\"display: none;\\">
        <div class=\\"m-stepper__content__wrapper\\" data-v-8231b20c=\\"\\">Content 3</div>
      </div>
    </transition-group-stub>
  </div>
</div>"
`;
